statsmodels/sandbox/gam_gsoc2015/tests/test_gam.py

```diff
@@ -0,0 +1,269 @@
+        hess = np.flipud(hess)
+        hess = np.fliplr(hess)
+        assert norm(hess - gam_der2)/25 < 1, 'error in the hessian of the GAM. Err=' + str(norm(hess - gam_der2)/25)
+    return
+
+
+def test_approximation():
+    x, y, basis, cov_der2, der2 = sample_data()
+    alpha = 0
+    for i in range(10):
+        params = np.random.randint(-2, 2, 5)
+        cost, err, itg = cost_function(params, basis, y, alpha)
+        gp = GamPenalty(alpha=alpha, cov_der2=cov_der2, der2=der2)
+        gam_itg = gp.func(params)
+        glm_gam = GLMGam(y, basis, penal = gp)
+        res_glm_gam = glm_gam.fit(maxiter=1) # TODO: can this fit be removed? It is useless. We just need the log likelihood
+        gam_loglike = glm_gam.loglike(params)
+        assert norm(gam_loglike - err) < 1.e-10, 'erron in the MSE part of the cost function'
+    return
+
+
+def test_gam_glm():
+
+    data_from_r = pd.read_csv('results/prediction_from_mgcv.csv')
+    # dataset used to train the R model
+    x = data_from_r.x
+    y = data_from_r.y
+
+    df = 10
+    degree = 5
+    basis, der_basis, der2 = make_bsplines_basis(x, df=df, degree=degree)
+    cov_der2 = np.dot(der2.T, der2)
+
+    # y_mgcv is obtained from R with the following code
+    # g = gam(y~s(x, k = 10, bs = "cr"), data = data, scale = 80)
+    y_mgcv = data_from_r.y_est
+
+    alpha = 0.05
+    gp = GamPenalty(alpha=alpha, cov_der2=cov_der2, der2=der2)
+    glm_gam = GLMGam(y, basis, penal=gp)
+    res_glm_gam = glm_gam.fit(maxiter=10000)
+    y_gam = np.dot(basis, res_glm_gam.params)
+
+    # plt.plot(x, y_gam, label='gam')
+    # plt.plot(x, y_mgcv, label='mgcv')
+    # plt.plot(x, y, '.', label='y')
+    # plt.legend()
+    # plt.show()
+
+    approx_error = norm(y_gam - y_mgcv)/len(y_mgcv)
+    assert approx_error < 0.01, 'The mean error between y_gam and y_mgcv is:' + str(approx_error)
+    return
+
+def test_gam_discrete():
+
+    data_from_r = pd.read_csv('results/prediction_from_mgcv.csv')
+    # dataset used to train the R model
+    x = data_from_r.x
+    y = data_from_r.ybin
+
+    df = 10
+    degree = 5
+    basis, der_basis, der2 = make_bsplines_basis(x, df=df, degree=degree)
+    cov_der2 = np.dot(der2.T, der2)
+
+    # y_mgcv is obtained from R with the following code
+    # g = gam(y~s(x, k = 10, bs = "cr"), data = data, scale = 80)
+    y_mgcv = data_from_r.ybin_est
+
+    alpha = 0.00002
+    gp = GamPenalty(alpha=alpha, cov_der2=cov_der2, der2=der2)
+    lg_gam = LogitGam(y, basis, penal=gp)
+    res_lg_gam = lg_gam.fit(maxiter=10000)
+    y_gam = np.dot(basis, res_lg_gam.params)
+    y_gam = sigmoid(y_gam)
+    y_mgcv = sigmoid(y_mgcv)
+
+    # plt.plot(x, y_gam, label='gam')
+    # plt.plot(x, y_mgcv, label='mgcv')
+    # plt.plot(x, y, '.', label='y')
+    # plt.ylim(-0.4, 1.4)
+    # plt.legend()
+    # plt.show()
+
+    approx_error = norm(y_gam - y_mgcv)/len(y_mgcv)
+    assert approx_error < 0.01, 'The mean error between y_gam and y_mgcv is:' + str(approx_error)
```

Comment:
use `numpy.testing.assert_allclose` for the unit tests, and either set `rtol` or `atol`, look at other test files for examples